### PR TITLE
ENH: Add defaultdict support for dtype in read_csv

### LINKED ISSUE #41574: ENH: support defaultdict in read_csv dtype parameter
I have a large csv file with 15k columns with non-obvious names, of which 14990 are floating point numbers. I'd like to load them as floats without `read_csv` having to spend the time divining their types.

`dtype` allows providing a `dict`, but making one with all the column names is tedious and not always possible. The obvious solution is to provide a `defaultdict`, with a default of `np.float32`, and including entries for the other column types. Unfortunately currently, the default is silently ignored by `read_csv`. Presumably `read_csv` is not directly querying the dictionary, but rather checking first whether an item is there. 

If this is not possible, it would be helpful to include a warning to the user, or at least some mention in the documentation, that defaultdict is not supported. It took me a long time to figure out why my floats weren't being treated as float32 and why read_csv was still trying to determine the types of these columns.

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #41574 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/tests/io/parser/dtypes/test_dtypes_basic.py [hunk 1]
```diff
@@ -335,3 +336,29 @@ def test_nullable_int_dtype(all_parsers, any_int_ea_dtype):
     )
     actual = parser.read_csv(StringIO(data), dtype=dtype)
     tm.assert_frame_equal(actual, expected)
+
+
+@pytest.mark.parametrize("default", ["float", "float64"])
+def test_dtypes_defaultdict(all_parsers, default):
```
Comment (mroeschke) [inline pandas/tests/io/parser/dtypes/test_dtypes_basic.py:342]: `default` doesn't look used in the test?

File: pandas/tests/io/parser/dtypes/test_dtypes_basic.py [hunk 2]
```diff
@@ -335,3 +336,29 @@ def test_nullable_int_dtype(all_parsers, any_int_ea_dtype):
     )
     actual = parser.read_csv(StringIO(data), dtype=dtype)
     tm.assert_frame_equal(actual, expected)
+
```
Comment (mroeschke) [inline pandas/tests/io/parser/dtypes/test_dtypes_basic.py:339]: Might be also good to (smoke) test the error if the default type isn't a valid pandas_type

File: pandas/tests/io/parser/dtypes/test_dtypes_basic.py (hunk 1 above)
Comment (phofl) [reply pandas/tests/io/parser/dtypes/test_dtypes_basic.py:342]: Thx, missed that

**Comment by jreback** [comment]
cool can you rebase

File: pandas/tests/io/parser/dtypes/test_dtypes_basic.py [hunk 3]
```diff
@@ -335,3 +336,40 @@ def test_nullable_int_dtype(all_parsers, any_int_ea_dtype):
...
+def test_dtypes_defaultdict_mangle_dup_cols(all_parsers):
+    # GH#41574
+    data = """a,b,a,b,b.1
+1,2,3,4,5
+"""
+    dtype = defaultdict(lambda: "float64", a="int64")
+    dtype["b.1"] = "int64"
+    parser = all_parsers
+    result = parser.read_csv(StringIO(data), dtype=dtype)
+    expected = DataFrame({"a": [1], "b": 2.0, "a.1": [3], "b.2": [4.0], "b.1": [5]})
```
Comment (attack68) [inline pandas/tests/io/parser/dtypes/test_dtypes_basic.py:363]: did you miss the [] on "b", looks different but dont think there is a reason

Comment (phofl) [reply pandas/tests/io/parser/dtypes/test_dtypes_basic.py:363]: Forgot it yes, but does not really make a difference. 

Added nevertheless to be more consistent

**Comment by jreback** [comment]
this looked fine, can you merge master @phofl

**Comment by phofl** [comment]
Done

**Review by mroeschke — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl very nice as always!

**Pull request merged**